File: LeetCodeSolutions/LeetCode_0063.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]
        if obstacleGrid[0][0]:
            return 0
        dp[1][1] = 1
        i, j = 1, 1
        # dp[i][j] = dp[i-1][j] + dp[i][j-1]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if i == 1 and j == 1:
                    continue
                if obstacleGrid[i - 1][j - 1]:
                    dp[i][j] = 0
                else:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1]

        return dp[-1][-1]

File: LeetCodeSolutions/LeetCode_0063.cpp
/*
 * @lc app=leetcode.cn id=63 lang=cpp
 *
 * [63] 不同路径 II
 */

// @lc code=start
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int row = obstacleGrid.size();
        if (row == 0) return 0;
        int col = obstacleGrid[0].size();
        vector<int> dp(col, 0);
        dp[0] = obstacleGrid[0][0] ? 0: 1;
        for (int i = 0; i < row; ++i)
            for(int j = 0 ; j < col; ++j){
                if(obstacleGrid[i][j]) {dp[j] = 0; continue;}
                dp[j] += j > 0 ? dp[j - 1] : 0;
            }
        return dp[col - 1];
    }
};
// @lc code=end

